public/app/features/explore/state/actions.ts

```diff
@@ -0,0 +1,836 @@
+  transaction: QueryTransaction;
+}
+
+export interface QueryTransactionSuccessAction {
+  type: ActionTypes.QueryTransactionSuccess;
+  exploreId: ExploreId;
+  history: HistoryItem[];
+  queryTransactions: QueryTransaction[];
+}
+
+export interface RemoveQueryRowAction {
+  type: ActionTypes.RemoveQueryRow;
+  exploreId: ExploreId;
+  index: number;
+}
+
+export interface ScanStartAction {
+  type: ActionTypes.ScanStart;
+  exploreId: ExploreId;
+  scanner: RangeScanner;
+}
+
+export interface ScanRangeAction {
+  type: ActionTypes.ScanRange;
+  exploreId: ExploreId;
+  range: RawTimeRange;
+}
+
+export interface ScanStopAction {
+  type: ActionTypes.ScanStop;
+  exploreId: ExploreId;
+}
+
+export interface StateSaveAction {
+  type: ActionTypes.StateSave;
+}
+
+export type Action =
+  | AddQueryRowAction
+  | ChangeQueryAction
+  | ChangeSizeAction
+  | ChangeTimeAction
+  | ClickClearAction
+  | ClickCloseSplitAction
+  | ClickExampleAction
+  | ClickGraphButtonAction
+  | ClickLogsButtonAction
+  | ClickSplitAction
+  | ClickTableButtonAction
+  | HighlightLogsExpressionAction
+  | InitializeExploreAction
+  | InitializeExploreSplitAction
+  | LoadDatasourceFailureAction
+  | LoadDatasourceMissingAction
+  | LoadDatasourcePendingAction
+  | LoadDatasourceSuccessAction
+  | ModifyQueriesAction
+  | QueryTransactionFailureAction
+  | QueryTransactionStartAction
+  | QueryTransactionSuccessAction
+  | RemoveQueryRowAction
+  | ScanRangeAction
+  | ScanStartAction
+  | ScanStopAction;
+type ThunkResult<R> = ThunkAction<R, StoreState, undefined, Action>;
+
+export function addQueryRow(exploreId: ExploreId, index: number): AddQueryRowAction {
+  const query = generateEmptyQuery(index + 1);
+  return { type: ActionTypes.AddQueryRow, exploreId, index, query };
+}
+
+export function changeDatasource(exploreId: ExploreId, datasource: string): ThunkResult<void> {
+  return async dispatch => {
+    const instance = await getDatasourceSrv().get(datasource);
+    dispatch(loadDatasource(exploreId, instance));
+  };
+}
+
+export function changeQuery(
+  exploreId: ExploreId,
+  query: DataQuery,
+  index: number,
+  override: boolean
+): ThunkResult<void> {
+  return dispatch => {
+    // Null query means reset
+    if (query === null) {
+      query = { ...generateEmptyQuery(index) };
+    }
+
+    dispatch({ type: ActionTypes.ChangeQuery, exploreId, query, index, override });
+    if (override) {
+      dispatch(runQueries(exploreId));
+    }
+  };
+}
+
+export function changeSize(
+  exploreId: ExploreId,
+  { height, width }: { height: number; width: number }
+): ChangeSizeAction {
+  return { type: ActionTypes.ChangeSize, exploreId, height, width };
+}
+
+export function changeTime(exploreId: ExploreId, range: TimeRange): ThunkResult<void> {
+  return dispatch => {
+    dispatch({ type: ActionTypes.ChangeTime, exploreId, range });
+    dispatch(runQueries(exploreId));
+  };
+}
+
+export function clickClear(exploreId: ExploreId): ThunkResult<void> {
```

Comment:
Talking about those actions that names start with `click...`. IMO they introduce an unnecessary level of indirectness into Explore's code. Take a look at the usage: 

```ts
 onClickClear = () => {
    this.props.clickClear(this.props.exploreId);
};
```

When I read this code the only thing I learn is that someone clicked clear. What I would like to learn though is what is the effect of the click, in this case this would probably be `clearAllQueries` or `clearQueriesAndPersistState` depending on how much information is really necessary.

Reply:
IMO handlers, actions and reducers are 3 concerns. But sometimes they overlap and this is where it gets messy. Reducers should be dumb, but not as dumb as to blindly apply only the state that they have been given (otherwise, what are they reducing?). Handler should translate to an action, because that's what enters the state logic. If the actions are too abstracted, they prescribe the state change (IMO they should not). Instead they should be reusable from an action point of view: "clear all" may be a button that occurs on the top bar, but also in a menu, etc.
For that concrete case I agree that the better name would be `clearAll`. But I'd resist on saying what it clears.

Reply:
I agree with some of this, the click part of the action names & function names seems unnecessary. Would prefer to have it be more domain oriented. CLEAR_EXPLORE_QUERIES. 
CLOSE_EXPLORE_SPLIT, or prefix the actions name constants with EXPLORE.

A lot of these action names needs to be changed to include EXPLORE as they are too generic and action names are global.